java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationState.java

```diff
@@ -31,15 +31,30 @@
   /**
    * This value means that the checked relation is fulfilled by the set of known relations
    */
-  FULFILLED(true, BooleanConstraint.FALSE),
+  FULFILLED(true, BooleanConstraint.FALSE) {
+    @Override
+    RelationState invert() {
+      return UNFULFILLED;
+    }
+  },
   /**
    * This value means that the checked relation is not fulfilled by the set of known relations
    */
-  UNFULFILLED(true, BooleanConstraint.TRUE),
+  UNFULFILLED(true, BooleanConstraint.TRUE) {
+    @Override
+    RelationState invert() {
+      return FULFILLED;
+    }
+  },
   /**
    * This value means that the checked relation is not determined by the set of known relations
    */
-  UNDETERMINED(false, null);
+  UNDETERMINED(false, null) {
+    @Override
+    RelationState invert() {
+      return UNDETERMINED;
+    }
+  };
 
   private final boolean determined;
```

Comment:
Thanks for suggestion, indeed it is simpler now. For `rejects` I also broke down the logic as in `invert` , so I removed all the fields and constructor